File: SimpleMazeBuilder1/SimpleMazeBuilder1/Grid.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SimpleMazeBuilder1
{
    public class Grid
    {
        public static String NEWLINE = "\r\n";

        public int Rows { get; set; }
        public int Columns { get; set; }

        private Cell[,] _cells;
        public Cell[,] Cells
        {
            get { return _cells; }
            set { _cells = value; }
        }

        public Grid(int rows, int cols)
        {
            this.Rows = rows;
            this.Columns = cols;

            _cells = new Cell[rows,cols];
            createCells();
            configureCells();
        }

        private void createCells()
        {
            int r, c = 0;

            for (r = 0; r < this.Rows; r++)
            {
                for (c = 0; c < this.Columns; c++)
                {
                    _cells[r,c] = new Cell(r, c);
                }
            }
        }

        private void configureCells()
        {
            int r,c = 0;

            for (r = 0; r < this.Rows; r++)
            {
                for (c = 0; c < this.Columns; c++)
                {
                    Cell cell = _cells[r, c];

                    //cell.Row = r;
                    //cell.Column = c;
                    // if the neighbor in a direction would be within the grid, add it, else null
                    if (r > 0) { cell.North = _cells[r - 1, c]; }
                    if (r < this.Rows - 1) { cell.South = _cells[r + 1, c]; }
                    if (c > 0) { cell.West = _cells[r, c - 1]; }
                    if (c < this.Columns - 1) { cell.East = _cells[r, c + 1]; }
                }
            }
            Console.WriteLine("Grid of " + this.Rows + " rows by " + this.Columns + " columns initialized.");
        }

        public Cell getRandomCell()
        {
            Random rand = new Random();
            int row = rand.Next(this.Rows);
            int col = rand.Next(this.Columns);

            return this.Cells[row, col];
        }

        public int getSize()
        {
            return this.Rows * this.Columns;
        }

        // Display as ASCII
        public String toString()
        {
            String output;
            String top, bottom;
            String eastBound, southBound, corner;
            String body = "   "; // 3 spaces
            Cell c;

            // Draw top of maze, no breaks
            output = "+";
            for (int j=0; j < this.Columns; j++)
            {
                output += "---+";
            }
            output += NEWLINE;

            // Draw each row
            for (int i = 0; i < this.Rows; i++)
            {
                top = "|";
                bottom = "+";
                for (int j = 0; j < this.Columns; j++)
                {
                    
                    c = this.Cells[i,j];
                    if (c.isLinked(c.East))
                    {
                        eastBound = " ";
                    }
                    else
                    {
                        eastBound = "|";
                    }
                    top = top + body + eastBound;

                    if (c.isLinked(c.South))
                    {
                        southBound = "   "; // 3 spaces
                    } 
                    else
                    {
                        southBound = "---";
                    }
                    corner = "+";
                    bottom = bottom + southBound + corner;
                }
                output += top + NEWLINE;
                output += bottom + NEWLINE;
            }


            Console.WriteLine(output);
            return output;
        }
        
    }
       
}

File: py_port/grid.py
import random
import sys
import os

# Add the current directory to the path so we can import modules
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from cell import Cell

class Grid:
    """
    Grid represents the maze as a collection of cells in a 2D grid.
    """
    # Direction offsets (row, col)
    DIRECTION_OFFSETS = {
        Cell.NORTH: (0, -1),
        Cell.SOUTH: (0, 1),
        Cell.EAST: (1, 0),
        Cell.WEST: (-1, 0)
    }
    
    def __init__(self, rows, cols):
        """Initialize a new grid with the given dimensions."""
        self.rows = rows
        self.cols = cols
        self.cells = self._initialize_cells()
    
    def _initialize_cells(self):
        """Create and configure cells for the grid."""
        # Create cells
        cells = [[Cell(r, c) for c in range(self.cols)] for r in range(self.rows)]
        return cells
    
    def is_valid(self, row, col):
        """Check if the given row and column are within the grid bounds."""
        return 0 <= row < self.rows and 0 <= col < self.cols
    
    def at(self, row, col):
        """Get the cell at the given row and column."""
        if not self.is_valid(row, col):
            raise IndexError(f"Cell position ({row}, {col}) is outside the grid")
        return self.cells[row][col]
    
    def random_int(self, min_val, max_val):
        """Generate a random integer between min_val and max_val (inclusive)."""
        return random.randint(min_val, max_val)
    
    def link_cells(self, row1, col1, direction):
        """Link the cell at (row1, col1) to its neighbor in the given direction."""
        if not self.is_valid(row1, col1):
            return
            
        row_offset, col_offset = Grid.DIRECTION_OFFSETS[direction]
        row2 = row1 + row_offset
        col2 = col1 + col_offset
        
        if not self.is_valid(row2, col2):
            return
            
        # Link both cells
        self.at(row1, col1).link(direction)
        self.at(row2, col2).link(Cell.OPPOSITES[direction])
    
    def display(self):
        """Display the maze as ASCII art."""
        # Display the top border
        output = ['+' + '---+' * self.cols]
        
        for r in range(self.rows):
            # Display cell contents and eastern boundaries
            row = ['|']
            eastern_boundary = ['+']
            
            for c in range(self.cols):
                cell = self.at(r, c)
                
                # Cell contents (3 spaces)
                row.append('   ')
                
                # Eastern boundary
                if c < self.cols - 1 and cell.linked(Cell.EAST):
                    row.append(' ')
                else:
                    row.append('|')
                
                # Southern boundary
                if r < self.rows - 1 and cell.linked(Cell.SOUTH):
                    eastern_boundary.append('   +')
                else:
                    eastern_boundary.append('---+')
            
            output.append(''.join(row))
            output.append(''.join(eastern_boundary))
        
        return '\n'.join(output)